src/prefs-dialog.c

```diff
@@ -353,200 +314,205 @@ sync_send_message_to_content (PrefsDialog *dialog,
-     * until the verification has finished. */
-    if (json_object_get_boolean_member (data, "verified") == FALSE) {
-      FxACallbackData *cb_data;
-
-      text = g_strdup_printf ("<span fgcolor='#e6780b'>%s</span>",
-                              _("Please don’t leave this page until you have completed the verification."));
-      gtk_label_set_markup (GTK_LABEL (dialog->sync_sign_in_details), text);
-      gtk_widget_set_visible (dialog->sync_sign_in_details, TRUE);
-
-      cb_data = fxa_callback_data_new (dialog, email, uid, sessionToken,
-                                       keyFetchToken, unwrapBKey, tokenID,
-                                       reqHMACkey, respHMACkey, respXORkey);
-      dialog->fxa_id = g_timeout_add_seconds (2, (GSourceFunc)poll_fxa_server, cb_data);
-
-      g_free (text);
-    } else {
-      char *bundle;
+    goto out_no_error;
+  }
+  if (g_strcmp0 (command, "fxaccounts:login")) {
+    g_warning ("Unexepected command: %s", command);
+    goto out_error;
+  }
 
-      bundle = ephy_sync_service_start_sign_in (service, tokenID, reqHMACkey);
-      ephy_sync_service_finish_sign_in (service, email, uid, sessionToken, keyFetchToken,
-                                        unwrapBKey, bundle, respHMACkey, respXORkey);
+  /* Login command. */
+  gtk_widget_set_visible (dialog->sync_firefox_iframe_label, FALSE);
 
-      g_free (bundle);
-    }
-  } else {
-    g_warning ("Unexepected command: %s", command);
+  data = json_object_get_object_member (message, "data");
+  if (!data) {
+    g_warning ("JSON object has invalid 'data' member");
+    goto out_error;
+  }
+  email = json_object_get_string_member (data, "email");
+  uid = json_object_get_string_member (data, "uid");
+  session_token = json_object_get_string_member (data, "sessionToken");
+  key_fetch_token = json_object_get_string_member (data, "keyFetchToken");
+  unwrap_b_key = json_object_get_string_member (data, "unwrapBKey");
+  if (!email || !uid || !session_token || !key_fetch_token || !unwrap_b_key) {
+    g_warning ("JSON object has missing or invalid members");
+    goto out_error;
   }
+  if (!json_object_has_member (data, "verified") ||
+      !JSON_NODE_HOLDS_VALUE (json_object_get_member (data, "verified"))) {
+    g_warning ("JSON object has missing or invalid 'verified' member");
+    goto out_error;
+  }
+
+  if (!json_object_get_boolean_member (data, "verified"))
+    sync_sign_in_details_show (dialog, _("Please don’t leave this page until "
+                                         "you have completed the verification."));
+  ephy_sync_service_do_sign_in (dialog->sync_service, email, uid,
+                                session_token, key_fetch_token, unwrap_b_key);
+  goto out_no_error;
 
-out:
+out_error:
+  sync_sign_in_details_show (dialog, _("Something went wrong, please try again."));
+  webkit_web_view_load_uri (dialog->fxa_web_view, FXA_IFRAME_URL);
+out_no_error:
+  if (node)
+    json_node_unref (node);
```

Comment:
@mcatanzaro This function has been simplified in the later commits and the memory allocation/deallocation is easier to follow there. There's no need for you to double-check, I made sure everything is freed correctly.

(`json` is a `JsonObject` owned by the parent `JsonNode`, so there's no need to free it there.)